src/js_native_api_types.h

```diff
@@ -154,4 +154,20 @@ typedef struct {
 } napi_type_tag;
 #endif  // NAPI_VERSION >= 8
 
+#ifdef NAPI_EXPERIMENTAL
+typedef enum {
+  node_api_native_data_flags_none = 0,
+  // may touch GC and must run in SetImmediate
+  node_api_native_data_flags_finalize_uses_js = 1 << 0
+} node_api_native_data_flags;
+
+typedef struct {
+  size_t version;  // version of the struct - must be 0.
+  void* data;
+  napi_finalize finalize_cb;
+  void* finalize_hint;
+  node_api_native_data_flags flags;
+} node_api_native_data;
```

Comment:
Where will add-ons store these structures? If they are stored on the heap, how will they be freed? Might it not be simpler to just not have this structure, and directly pass in its members as parameters to the new APIs? Yes, it would make the signatures more verbose, but in exchange it might remove a memory management burden from the add-on maintainers.